server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -226,61 +279,139 @@ public float getShardBalance() {
+
+        public Balancer(
+            WriteLoadForecaster writeLoadForecaster,
+            ClusterInfoService clusterInfoService,
+            RoutingAllocation allocation,
+            WeightFunction weight,
+            float threshold
+        ) {
+            this.writeLoadForecaster = writeLoadForecaster;
             this.allocation = allocation;
             this.weight = weight;
             this.threshold = threshold;
             this.routingNodes = allocation.routingNodes();
             this.metadata = allocation.metadata();
+            this.clusterInfoService = clusterInfoService;
             avgShardsPerNode = ((float) metadata.getTotalNumberOfShards()) / routingNodes.size();
+            avgWriteLoadPerNode = getTotalWriteLoad(writeLoadForecaster, metadata) / routingNodes.size();
+            avgDiskUsageInBytesPerNode = ((double) getTotalDiskUsageInBytes(clusterInfoService, metadata) / routingNodes.size());
             nodes = Collections.unmodifiableMap(buildModelFromAssigned());
             sorter = newNodeSorter();
         }
 
+        private static double getTotalWriteLoad(WriteLoadForecaster writeLoadForecaster, Metadata metadata) {
+            double writeLoad = 0.0;
+            for (IndexMetadata indexMetadata : metadata.indices().values()) {
+                writeLoad += getIndexWriteLoad(writeLoadForecaster, indexMetadata);
+            }
+            return writeLoad;
+        }
+
+        private static double getIndexWriteLoad(WriteLoadForecaster writeLoadForecaster, IndexMetadata indexMetadata) {
+            var shardWriteLoad = writeLoadForecaster.getForecastedWriteLoad(indexMetadata).orElse(0.0);
+            return shardWriteLoad * numberOfCopies(indexMetadata);
+        }
+
+        private static long getTotalDiskUsageInBytes(ClusterInfoService clusterInfoService, Metadata metadata) {
+            long totalDiskUsageInBytes = 0;
+            for (IndexMetadata indexMetadata : metadata.indices().values()) {
+                totalDiskUsageInBytes += getIndexDiskUsageInBytes(clusterInfoService, indexMetadata);
+            }
+            return totalDiskUsageInBytes;
+        }
+
+        private static long getIndexDiskUsageInBytes(ClusterInfoService clusterInfoService, IndexMetadata indexMetadata) {
+            OptionalLong forecastedShardSizeInBytes = indexMetadata.getForecastedShardSizeInBytes();
+            final long indexDiskUsageInBytes;
+            if (forecastedShardSizeInBytes.isPresent()) {
+                indexDiskUsageInBytes = forecastedShardSizeInBytes.getAsLong() * numberOfCopies(indexMetadata);
+            } else {
+                long totalSizeInBytes = 0;
+                int shardCount = 0;
+                ClusterInfo clusterInfo = clusterInfoService.getClusterInfo();
+                for (int shardId = 0; shardId < indexMetadata.getNumberOfShards(); shardId++) {
+                    for (Boolean primary : List.of(true, false)) {
```

Comment:
should this one iterate over `0..indexMetadata.getNumberOfReplicas()+1` ?

Reply:
oh right 🤦

Reply:
Tackled in https://github.com/elastic/elasticsearch/pull/91603/commits/0d857bb776787a9dfdc4517bc2b213ae3bc103ec